lib/transfer.c

```diff
@@ -918,13 +918,42 @@ static CURLcode readwrite_upload(struct Curl_easy *data,
 
   do {
     curl_off_t nbody;
+    ssize_t offset = 0;
+
+    if(0 != k->upload_present && k->upload_present < 1024 && !k->upload_done) {
+      DEBUGASSERT(data->set.upload_buffer_size >= 1024);
```

Comment:
I would prefer a define instead of the hard-coded number 1024 here. Then you can add a comment where the threshold is defined explaining the purpose of this value.

The buffer size is changeable in the API. Should the threshold perhaps rather be a share of that buffer size?

Reply:
Ok.  Where would you like the `#define` located?  In the func?  Above the func?  In a header?

The purpose is to avoid less efficient tiny writes.  The value is set to limit the size and overhead of shifting data at the end of the buffer to the beginning of the buffer.  (The overhead of copying is low compared to inefficient writes to the network, but gets slightly more expensive as the buffer size increases and more data is copied.)

If you prefer a share of the buffer, then I propose 1/32 of the buffer size (upload_buffer_size >> 5).  The reason I had chosen 1024 is that the HTTP/2 default max frame size is 16k.  However, you rightly point out that this code is used by many different protocols, not just HTTP/2.

Just about any reasonable value should be better than the current behavior, so please choose something and let me know your preference.

Reply:
modified to `(data->set.upload_buffer_size >> 5)` in the updated patch.  Will change to whatever your preference.

Reply:
I honestly think, that a plain and hard-coded `1024` does the job better. I only figured it's purpose out, after seeing this conversation and good code should explain itself.

Reply:
@emilengler The block immediately following and a few lines below where `offset` is assigned states: `/* only read more data if there's no upload data already present in the upload buffer, or if appending to upload buffer */`.  I agree that good code should explain itself, often without comments.  At the same time, hard-coded magic numbers such as 1024 often do not add anything to clarity, which is why @bagder requested a better-named `#define`.  I therefore disagree with your comment "`1024` does the job better.".  It does not do the job better.  I am awaiting feedback from @bagder if he prefers `#define foo 1024` or `#define foo(data) ((data)->set.upload_buffer_size >> 5)`, and what 'foo' should be named and where the `#define` should be located in a .c or .h.

Reply:
I created a `#define curl_upload_refill_watermark(data)`